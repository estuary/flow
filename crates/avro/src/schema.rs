use super::{FLOW_EXTRA_NAME, FLOW_KEY_NAME};
use apache_avro::schema as avro;
use json::schema::formats::Format;
use json::schema::types;
use std::fmt::Write;

/// Map a Shape at the given location into an AVRO schema.
/// If the location is not required and has no default, it may implicitly be none.
pub fn shape_to_avro(loc: json::Location, shape: doc::Shape, required: bool) -> avro::Schema {
    let mut type_ = shape.type_;

    // Is this location nullable ? NULL may union with any other schema.
    let nullable = if shape.type_.overlaps(types::NULL) || (!required && shape.default.is_none()) {
        type_ = type_ - types::NULL;
        true
    } else {
        false
    };

    let base = if type_.overlaps(types::STRING) {
        match (type_ - types::STRING, shape.string.format) {
            // If the location may contain a fractional number, map to a double.
            // This could lose precision of a string number, but we're already using f64's,
            // and these are typically representing Nan or Inf.
            (types::INT_OR_FRAC | types::FRACTIONAL, Some(Format::Number | Format::Integer)) => {
                avro::Schema::Double
            }
            // Or, if the location may contain an integer, map to a string.
            // A string integer may not fit in a 64-bit signed integer.
            (types::INTEGER, Some(Format::Number | Format::Integer)) => avro::Schema::String,
            // String which holds an RFC-3339 date.
            (types::INVALID, Some(Format::Date)) => avro::Schema::Date,
            // String which holds an RFC-3339 date-time.
            (types::INVALID, Some(Format::DateTime)) => avro::Schema::TimestampMicros,
            // String which holds a canonical UUID.
            (types::INVALID, Some(Format::Uuid)) => avro::Schema::Uuid,
            // String which holds an ISO-8601 duration.
            (types::INVALID, Some(Format::Duration)) => avro::Schema::Duration,
            // A regular string.
            (types::INVALID, _) => avro::Schema::String,
            // Other combinations fall back to JSON encoding.
            _ => raw_json_schema(loc),
        }
    } else {
        match type_ {
            types::ARRAY => array_to_avro(loc, shape.array),
            types::BOOLEAN => avro::Schema::Boolean,
            types::INTEGER => avro::Schema::Long,
            types::INT_OR_FRAC | types::FRACTIONAL => avro::Schema::Double,
            types::OBJECT => object_to_avro(loc, shape.object),
            // Other combinations fall back to JSON encoding.
            _ => raw_json_schema(loc),
        }
    };

    if !nullable || matches!(base, avro::Schema::Null) {
        base
    } else {
        avro::Schema::Union(avro::UnionSchema::new(vec![base, avro::Schema::Null]).unwrap())
    }
}

// Map a key extracted from a Shape into a flattened Avro Record.
pub fn key_to_avro(key: &[doc::Pointer], shape: doc::Shape) -> avro::Schema {
    let loc_root = json::Location::Root;
    let loc_key = loc_root.push_prop("Key");
    let loc_parts = loc_key.push_prop("Parts");

    let mut parts = Vec::new();

    // Map each key component into a field with a structured name.
    for (position, ptr) in key.iter().enumerate() {
        let name = format!("p{}", position + 1);
        let (shape, _) = shape.locate(ptr);

        let default = shape.default.as_ref().map(|d| d.0.clone());
        let schema = shape_to_avro(loc_parts.push_prop(&name), shape.clone(), true);

        parts.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default,
            doc: None,
            name,
            order: avro::RecordFieldOrder::Ascending,
            position,
            schema,
        });
    }

    // Wrap the record of key components in a single-field parent record.
    let parts_schema = avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc_parts),
        aliases: None,
        doc: None,
        fields: parts,
        attributes: Default::default(),
        lookup: Default::default(),
    });

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc_key),
        aliases: None,
        doc: None,
        fields: vec![avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default: None,
            doc: None,
            name: FLOW_KEY_NAME.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: 0,
            schema: parts_schema,
        }],
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map an Object Shape into an Avro schema (a Record or Map).
fn object_to_avro(loc: json::Location, obj: doc::shape::ObjShape) -> avro::Schema {
    let mut fields: Vec<avro::RecordField> = Vec::new();
    let mut extra = doc::Shape::nothing();

    for pattern in obj.pattern_properties {
        extra = doc::Shape::union(extra, pattern.shape);
    }
    if let Some(addl) = obj.additional_properties {
        extra = doc::Shape::union(extra, *addl);
    } else {
        extra = doc::Shape::union(extra, doc::Shape::anything());
    }

    // If there are no explicit properties, but this object may have pattern or
    // additional properties, then interpret it as an Avro map.
    if extra.type_ != types::INVALID && obj.properties.is_empty() {
        let schema = shape_to_avro(loc, extra, true);
        return avro::Schema::map(schema);
    }

    // Otherwise, build a Record which may have a placeholder
    // field for additional dynamic properties.

    for prop in obj.properties {
        if !AVRO_FIELD_RE.is_match(&prop.name) {
            extra = doc::Shape::union(extra, prop.shape);
            continue; // Cannot be represented under Avro's name restrictions.
        }
        let default = prop.shape.default.as_ref().map(|d| d.0.clone());
        let schema = shape_to_avro(loc.push_prop(&prop.name), prop.shape, prop.is_required);

        fields.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default,
            doc: None,
            name: prop.name.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: fields.len(),
            schema,
        })
    }

    // If extra dynamic properties are possible, add a trailing field for them.
    // This field MUST appear last in the record schema.

    if extra.type_ != types::INVALID {
        let schema = shape_to_avro(loc.push_prop(FLOW_EXTRA_NAME), extra, true);
        let schema = avro::Schema::map(schema);

        fields.push(avro::RecordField {
            aliases: None,
            custom_attributes: Default::default(),
            default: None,
            doc: None,
            name: FLOW_EXTRA_NAME.to_string(),
            order: avro::RecordFieldOrder::Ascending,
            position: fields.len(),
            schema,
        })
    }

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc),
        aliases: None,
        doc: None,
        fields,
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map an Array Shape into an Avro Array.
fn array_to_avro(loc: json::Location, shape: doc::shape::ArrayShape) -> avro::Schema {
    let mut items = doc::Shape::nothing();

    if let Some(addl) = shape.additional_items {
        items = doc::Shape::union(items, *addl);
    } else if !matches!(shape.max_items, Some(m) if m as usize <= shape.tuple.len()) {
        items = doc::Shape::union(items, doc::Shape::anything());
    }

    for shape in shape.tuple {
        items = doc::Shape::union(items, shape);
    }

    let items = shape_to_avro(loc.push_prop("_items"), items, true);
    avro::Schema::array(items)
}

// Map a location into a special Schema holding a string-encoded JSON value.
fn raw_json_schema(loc: json::Location) -> avro::Schema {
    let fields = vec![avro::RecordField {
        aliases: None,
        custom_attributes: Default::default(),
        default: None,
        doc: None,
        name: "json".to_string(),
        order: avro::RecordFieldOrder::Ignore,
        position: 0,
        schema: avro::Schema::String,
    }];

    avro::Schema::Record(avro::RecordSchema {
        name: location_to_name(loc.push_prop("RawJSON")),
        aliases: None,
        doc: None,
        fields,
        attributes: Default::default(),
        lookup: Default::default(),
    })
}

// Map a Location into an Avro dot-separated namespace and name.
fn location_to_name(loc: json::Location) -> avro::Name {
    let name = loc.fold(String::new(), move |loc, mut n: String| {
        match loc {
            json::Location::Root => n.push_str("root"),
            json::Location::Property(json::LocatedProperty { name, .. }) => {
                n.push('.');
                n.push_str(name);
            }
            json::Location::Item(json::LocatedItem { index, .. }) => {
                write!(&mut n, ".{index}").unwrap();
            }
            _ => unreachable!(),
        };
        n
    });

    avro::Name::new(&name).unwrap()
}

lazy_static::lazy_static! {
    // The set of allowed characters in an AVRO field name.
    pub static ref AVRO_FIELD_RE : regex::Regex = regex::Regex::new(r"^[A-Za-z_][A-Za-z0-9_]*$").unwrap();
}

#[cfg(test)]
mod test {
    use serde_json::json;

    #[test]
    fn test_complex_schema() {
        let fixture = json!({
          "type": "object",
          "properties": {
            "a_bool": {"type": "boolean"},
            "a_date": {"type": "string", "format": "date"},
            "a_date_time": {"type": "string", "format": "date-time"},
            "a_duration": {"type": "string", "format": "duration"},
            "a_fractional": {"type": "number", "not": {"type": "integer"}},
            "a_integer_like": {"oneOf": [{"type": "integer"}, {"type": "string", "format": "integer"}]},
            "a_null": {"type": "null"},
            "a_number_like":  {"oneOf": [{"type": "number"}, {"type": "string", "format": "number"}]},
            "a_number_like_mixed1": {"oneOf": [{"type": "integer"}, {"type": "string", "format": "number"}]},
            "a_number_like_mixed2": {"oneOf": [{"type": "number"}, {"type": "string", "format": "integer"}]},
            "a_str": {"type": "string", "default": "xyz"},
            "an_int": {"type": "integer", "default": 32},
            "arr_items": {"type": "array", "items": {"type": "boolean"}},
            "arr_tuple": {
              "type": "array",
              "items": [{"type": "number"}, {"type": "string", "format": "number"}],
              "maxItems": 2,
            },
            "arr_tuple_items": {
              "type": "array",
              "items": [{"type": "integer"}, {"type": "integer"}],
              "additionalItems": {"type": "number"},
            },
            "map": {
                "type": "object",
                "additionalProperties": {"type": "boolean"},
            },
            "obj": {
              "type": "object",
              "properties": {
                "bar": {
                  "type": "object",
                  "properties": {"a_const": {"const": 42}, "filtered invalid avro name": {"const": 52}},
                  "additionalProperties": {"type": "integer"},
                  "patternProperties": {
                    "str_.*": {"type": "string", "format": "integer"},
                  },
                  "required": ["a_const"],
                },
                "foo": {"type": "boolean"},
              },
              "required": ["foo", "bar"],
              "additionalProperties": false,
            },
          },
          "required": ["an_int", "a_bool", "a_date", "obj"],
        }).to_string();

        let key = &["/a_bool", "/obj/a_map/a_const"];
        let key: Vec<_> = key.iter().map(|p| doc::Pointer::from_str(p)).collect();
        insta::assert_json_snapshot!(schema_test(&fixture, &key));
    }

    fn schema_test(json_schema: &str, key: &[doc::Pointer]) -> serde_json::Value {
        let (key, value) = crate::json_schema_to_avro(json_schema, key).unwrap();

        json!({
            "key": &key,
            "value": &value,
        })
    }
}
