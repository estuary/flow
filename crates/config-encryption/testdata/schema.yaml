type: object
properties:
  supa_secret:
    secret: true
    type: string
  supda_dupa_secret:
    airbyte_secret: true
    type: integer
  not_secret:
    type: string

