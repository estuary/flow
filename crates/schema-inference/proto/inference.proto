syntax = "proto3";

package schema_inference;

message InferenceRequest {
    string flow_collection = 1;
}

message InferenceResponse {
    oneof body {
        InferredSchema schema = 1;
    }
}

message InferredSchema {
    string schema_json = 1;
    uint64 documents_read = 2;
}

service InferenceService {
    rpc InferSchema (InferenceRequest) returns (InferenceResponse);
}