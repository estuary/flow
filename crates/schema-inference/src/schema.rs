use std::collections::{BTreeMap, BTreeSet};

use doc::inference::Shape;
use json::schema::types::{self, Set};
use schemars::gen::SchemaGenerator;
use schemars::schema::{InstanceType, RootSchema, Schema, SchemaObject, SingleOrVec};

use crate::shape;

#[derive(Debug, Default)]
pub struct SchemaBuilder {
    shape: Shape,
}

impl SchemaBuilder {
    pub fn new(shape: Shape) -> Self {
        Self { shape }
    }

    pub fn root_schema(&self) -> RootSchema {
        RootSchema {
            schema: to_schema(&self.shape).into_object(),
            meta_schema: SchemaGenerator::default().settings().meta_schema.clone(),
            ..Default::default()
        }
    }

    pub fn merge(self, other: Self) -> Self {
        Self {
            shape: shape::merge(self.shape, other.shape),
        }
    }
}

pub fn to_schema(shape: &Shape) -> Schema {
    let mut schema_object = SchemaObject {
        instance_type: Some(shape_type_to_schema_type(shape.type_)),
        format: shape.string.format.map(|f| f.to_string()),
        ..Default::default()
    };

    if shape.type_.overlaps(types::OBJECT) {
        let mut prop_schemas = BTreeMap::new();
        let mut required = BTreeSet::new();
        for obj_prop in shape.object.properties.iter() {
            prop_schemas.insert(obj_prop.name.clone(), to_schema(&obj_prop.shape));
            if obj_prop.is_required {
                required.insert(obj_prop.name.clone());
            }
        }
        let object = &mut schema_object.object();
        object.properties = prop_schemas;
        object.required = required;
    }

    if shape.type_.overlaps(types::ARRAY) {
        let mut array_items = Vec::new();
        for item in shape.array.tuple.iter() {
            array_items.push(to_schema(item));
        }
        schema_object.array().items = Some(flatten(array_items));
    }

    Schema::Object(schema_object)
}

pub fn shape_type_to_schema_type(type_set: Set) -> SingleOrVec<InstanceType> {
    let instance_types = type_set
        .iter()
        .map(parse_instance_type)
        .collect::<Vec<InstanceType>>();

    flatten(instance_types)
}

fn parse_instance_type(input: &str) -> InstanceType {
    match input {
        "array" => InstanceType::Array,
        "boolean" => InstanceType::Boolean,
        "fractional" => InstanceType::Number,
        "integer" => InstanceType::Integer,
        "null" => InstanceType::Null,
        "number" => InstanceType::Number,
        "object" => InstanceType::Object,
        "string" => InstanceType::String,
        other => panic!("unexpected type: {}", other),
    }
}

fn flatten<T>(mut vec: Vec<T>) -> SingleOrVec<T> {
    if vec.len() == 1 {
        SingleOrVec::Single(Box::new(vec.pop().unwrap()))
    } else {
        SingleOrVec::Vec(vec)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn simple_type_conversions() {
        fn assert_equiv(expected: InstanceType, actual: types::Set) {
            assert_eq!(
                SingleOrVec::Single(Box::new(expected)),
                shape_type_to_schema_type(actual)
            );
        }

        assert_equiv(InstanceType::Array, types::ARRAY);
        assert_equiv(InstanceType::Boolean, types::BOOLEAN);
        assert_equiv(InstanceType::Integer, types::INTEGER);
        assert_equiv(InstanceType::Null, types::NULL);
        assert_equiv(InstanceType::Number, types::INT_OR_FRAC);
        assert_equiv(InstanceType::Object, types::OBJECT);
        assert_equiv(InstanceType::String, types::STRING);
    }
}
