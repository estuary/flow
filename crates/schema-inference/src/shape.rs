use doc::inference::{ArrayShape, ObjProperty, ObjShape, Shape, StringShape};
use itertools::{EitherOrBoth, Itertools};
use json::schema::types;

// TODO(johnny): We are *very* close to being able to define merge()
// in terms of a trivial Shape::union(). The one reason we cannot,
// is because we would have to mark object shapes with additionalProperties: false,
// which is definitely something we should do but not until we have a tight
// feedback / re-publication cycle around inferred-schema violations.
//
// See comment on infer_object_shape()
//
// For now, the difference between merge() and union() is that merge
// does not prune properties present in one schema but not the other,
// where the other schema does not specifically restrict that property
// from existing. merge() also doesn't handle a slew of cases that
// union() *does* handle, and we should seek to remove it.
pub fn merge(lhs: Shape, rhs: Shape) -> Shape {
    let string = match (
        lhs.type_.overlaps(types::STRING),
        rhs.type_.overlaps(types::STRING),
    ) {
        (true, true) => StringShape::union(lhs.string, rhs.string),
        (_, false) => lhs.string,
        (false, true) => rhs.string,
    };
    let array = match (
        lhs.type_.overlaps(types::ARRAY),
        rhs.type_.overlaps(types::ARRAY),
    ) {
        (true, true) => merge_array_shapes(lhs.array, rhs.array),
        (_, false) => lhs.array,
        (false, true) => rhs.array,
    };
    let object = match (
        lhs.type_.overlaps(types::OBJECT),
        rhs.type_.overlaps(types::OBJECT),
    ) {
        (true, true) => merge_obj_shapes(lhs.object, rhs.object),
        (_, false) => lhs.object,
        (false, true) => rhs.object,
    };

    Shape {
        type_: lhs.type_ | rhs.type_,
        array,
        object,
        string,
        ..Default::default()
    }
}

fn merge_obj_shapes(lhs: ObjShape, rhs: ObjShape) -> ObjShape {
    // NOTE(johnny): This is an incorrect re-implementation of ObjShape::union(),
    // which omits a slew of important details.
    //
    // The one salient difference that we currently rely on is that
    // this implementation preserves properties in all cases.
    // ObjShape::union(), conversely, will "preserve" a property only if the other
    // side constrains that property to not exist (as otherwise its schema must be
    // widened to the point of being completely unconstrained,
    // which is why ObjShape::union() removes it).

    let properties = itertools::merge_join_by(
        lhs.properties.into_iter(),
        rhs.properties.into_iter(),
        |l, r| Ord::cmp(&l.name, &r.name),
    )
    .map(|eob| match eob {
        EitherOrBoth::Both(l, r) => ObjProperty {
            name: l.name,
            is_required: l.is_required && r.is_required,
            shape: merge(l.shape, r.shape),
        },
        EitherOrBoth::Left(l) => ObjProperty {
            name: l.name,
            is_required: false,
            shape: l.shape,
        },
        EitherOrBoth::Right(r) => ObjProperty {
            name: r.name,
            is_required: false,
            shape: r.shape,
        },
    })
    .collect::<Vec<ObjProperty>>();

    ObjShape {
        properties,
        patterns: vec![],
        additional: None,
    }
}

fn merge_array_shapes(lhs: ArrayShape, rhs: ArrayShape) -> ArrayShape {
    // NOTE(johnny): This is an incorrect re-implementation of ArrayShape::union(),
    // which exists only so that merge_obj_shapes() is called to handled
    // nested sub-objects.
    let tuple = lhs
        .tuple
        .into_iter()
        .zip_longest(rhs.tuple.into_iter())
        .map(|eob| match eob {
            EitherOrBoth::Both(l, r) => merge(l, r),
            EitherOrBoth::Left(l) => l,
            EitherOrBoth::Right(r) => r,
        })
        .collect::<Vec<Shape>>();

    ArrayShape {
        min: None,
        max: None,
        tuple,
        additional: None,
    }
}
