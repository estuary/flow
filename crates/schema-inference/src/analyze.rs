use schemars::schema::RootSchema;
use serde_json::Value as JsonValue;
use std::io::BufRead;

use crate::inference::infer_shape;
use crate::schema::SchemaBuilder;
use crate::shape;

type StreamResult = serde_json::Result<JsonValue>;

pub fn infer_schema<R: BufRead + 'static>(reader: R) -> Result<RootSchema, anyhow::Error> {
    let stream = serde_json::de::Deserializer::from_reader(reader).into_iter();
    let documents = stream;
    let schema = analyze(documents)?;

    Ok(schema)
}

fn analyze<S>(values: S) -> anyhow::Result<RootSchema>
where
    S: Iterator<Item = StreamResult>,
{
    let mut acc = None;

    for result in values {
        let shape = infer_shape(&result?);
        if let Some(acc_shape) = acc {
            acc = Some(shape::merge(acc_shape, shape));
        } else {
            acc = Some(shape);
        }
    }

    if let Some(shape) = acc {
        Ok(SchemaBuilder::new(shape).root_schema())
    } else {
        Err(anyhow::anyhow!("no documents found"))
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use serde_json::json;

    fn to_stream(documents: Vec<JsonValue>) -> impl Iterator<Item = StreamResult> {
        Box::new(documents.into_iter().map(Result::Ok))
    }

    #[test]
    fn test_generator_with_multiple_values() {
        let data = json!({"a_null_value": null, "boolean": true, "number": 123, "string": "else"});

        let schema = analyze(to_stream(vec![data])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn test_nested_values() {
        let data = json!({
            "id": 1,
            "title": "First Post",
            "author": {
                "name": "Bruce Wayne",
                "contact": "bruce@wayneenterprises.com",
            },
            "tags": ["meta"],
        });

        let schema = analyze(to_stream(vec![data])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn test_very_nested() {
        let data = json!({
            "one": {
                "two": {
                    "three": {
                        "four": {
                            "five": {
                                "f": "foo",
                            },
                            "e": false,
                        },
                        "d": null,
                    },
                    "c": 2.9999_f64,
                },
                "b": 2_u8,
            },
            "a": true,
        });

        let schema = analyze(to_stream(vec![data])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn simple_schema_merging() {
        let data_one = json!({
            "a": 1,
            "b": [true, false, true],
            "c": {
                "inner": null,
                "optional": []
            },
            "s": {
                "sometimes": "an object",
            },
            "x": "extra",
        });
        let data_two = json!({
            "a": 2,
            "b": null,
            "c": {
                "inner": "nested data",
            },
            "s": "sometimes a string",
            "y": ["why not"],
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn multi_merging() {
        let data_one = json!({
            "a": 1,
            "b": [{"truthful": "yes"}, {"truthful": "no", "harmless": "yes"}],
            "c": {
                "inner": null,
                "optional": []
            },
            "x": "extra",
        });
        let data_two = json!({
            "a": 2,
            "b": [true],
            "c": {
                "inner": "nested data",
            },
            "y": ["why not"],
        });
        let data_three = json!({
            "a": { "real": 2.5, "imaginary": -1 },
            "b": [{"truthful": true, "hurtful": true}, {"truthful": false}],
            "c": {
                "inner": { "details": "fascinating" },
                "optional": [4,5,6],
                "even_more_optional": true,
            },
            "x": "extra extra",
            "z": null,
        });

        let schema = analyze(to_stream(vec![data_one, data_two, data_three])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn string_format_integer_merging() {
        let data_one = json!({
            "a": 1,
        });
        let data_two = json!({
            "a": "2",
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "1",
        });
        let data_two = json!({
            "a": 2,
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn string_format_number_merging() {
        let data_one = json!({
            "a": 1,
        });
        let data_two = json!({
            "a": "2.1",
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": 1.1,
        });
        let data_two = json!({
            "a": "2",
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "1",
        });
        let data_two = json!({
            "a": 2.1,
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "1.1",
        });
        let data_two = json!({
            "a": 2,
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn string_format_number_back_to_string() {
        let data_one = json!({
            "a": 1,
        });
        let data_two = json!({
            "a": "test"
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "a",
        });
        let data_two = json!({
            "a": "1"
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);
    }

    #[test]
    fn string_format_number_special_strings() {
        let data_one = json!({
            "a": 1,
        });
        let data_two = json!({
            "a": "NaN"
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "Infinity",
        });
        let data_two = json!({
            "a": "1"
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);

        let data_one = json!({
            "a": "-Infinity",
        });
        let data_two = json!({
            "a": "1"
        });

        let schema = analyze(to_stream(vec![data_one, data_two])).unwrap();
        insta::assert_json_snapshot!(schema);
    }
}
