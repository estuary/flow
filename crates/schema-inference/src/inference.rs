use crate::shape;
use doc::inference::{ArrayShape, ObjProperty, ObjShape, Shape, StringShape};
use json::schema::{formats::Format, types};
use serde_json::Value as JsonValue;

pub fn infer_shape(value: &JsonValue) -> Shape {
    let mut shape = Shape {
        type_: types::Set::for_value(value),
        ..Default::default()
    };

    if let JsonValue::String(value) = value {
        shape.string = infer_string_shape(value);
    } else if let JsonValue::Array(inner) = value {
        shape.array = infer_array_shape(inner);
    } else if let JsonValue::Object(values) = value {
        shape.object = infer_object_shape(values);
    }

    shape
}

fn infer_string_shape(value: &str) -> StringShape {
    let format = match value {
        _ if Format::Integer.validate(value).is_ok() => Some(Format::Integer),
        _ if Format::Number.validate(value).is_ok() => Some(Format::Number),
        _ if Format::DateTime.validate(value).is_ok() => Some(Format::DateTime),
        _ if Format::Date.validate(value).is_ok() => Some(Format::Date),
        _ if Format::Uuid.validate(value).is_ok() => Some(Format::Uuid),
        _ => None,
    };

    StringShape {
        format,
        ..Default::default()
    }
}

fn infer_array_shape(inner: &[JsonValue]) -> ArrayShape {
    if let Some(shape) = inner
        .iter()
        .map(infer_shape)
        .reduce(|acc, v| shape::merge(acc, v))
    {
        ArrayShape {
            tuple: vec![shape],
            ..Default::default()
        }
    } else {
        Default::default()
    }
}

fn infer_object_shape(inner: &serde_json::Map<String, JsonValue>) -> ObjShape {
    let properties = inner
        .iter()
        .map(|(key, value)| ObjProperty {
            name: key.to_owned(),
            // TODO(johnny): Mark `required` once we have a tighter ability
            // to quickly update inferred collection schemas upon a violation.
            is_required: false,
            shape: infer_shape(value),
        })
        .collect();

    ObjShape {
        properties,
        // TODO(johnny): Once we get good at updating inferred schemas on violations,
        // we want to enable additionalProperties: false.
        // This does two things:
        //  1) It allows us to define schema inference strictly in terms of
        //     doc::inference::Shape::union(), since union will preserve properties
        //     on one side and not the other, so long as the other side is constrained
        //     to not exist. This is the *only* reason we define a separate "merge"
        //     operation in this crate.
        //  2) It ensures that, when new properties are added, we first propagate their
        //     projections to downstream materializations *before* we process the first
        //     such document. This is a significant UX improvement for users because
        //     added properties will dynamically be added to materializations, with
        //     all values immediately represented in the ongoing materialization.
        /*
        additional: Some(Box::new(Shape {
            type_: types::INVALID,
            ..Default::default()
        })),
        */
        ..Default::default()
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use json::schema::types;
    use serde_json::json;

    #[test]
    fn build_primitive_types() {
        let shape = infer_shape(&json!(true));
        assert!(shape.type_.is_single_scalar_type());
        assert_eq!(types::BOOLEAN, shape.type_);

        let shape = infer_shape(&json!("string".to_string()));
        assert!(shape.type_.is_single_scalar_type());
        assert_eq!(types::STRING, shape.type_);

        let shape = infer_shape(&json!(123));
        assert!(shape.type_.is_single_scalar_type());
        assert_eq!(types::INTEGER, shape.type_);

        let shape = infer_shape(&json!(null));
        assert_eq!(types::NULL, shape.type_);
    }

    #[test]
    fn build_string_types() {
        let shape = infer_shape(&json!("1"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::Integer), shape.string.format);

        let shape = infer_shape(&json!("100.00"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::Integer), shape.string.format);

        let shape = infer_shape(&json!("1.001"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::Number), shape.string.format);

        for t in ["NaN", "Infinity", "-Infinity"] {
            let shape = infer_shape(&json!(t));
            assert_eq!(shape.type_, types::STRING);
            assert_eq!(Some(Format::Number), shape.string.format);
        }

        let shape = infer_shape(&json!("2021-07-08T12:34:56.523Z"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::DateTime), shape.string.format);

        let shape = infer_shape(&json!("2021-07-08"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::Date), shape.string.format);

        let shape = infer_shape(&json!("34c1506a-a0da-498e-b690-ea5183026979"));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(Some(Format::Uuid), shape.string.format);

        let shape = infer_shape(&json!(
            "2021-07-08 plus 34c1506a-a0da-498e-b690-ea5183026979 12.34"
        ));
        assert_eq!(shape.type_, types::STRING);
        assert_eq!(None, shape.string.format);
    }

    #[test]
    fn build_array_types() {
        let shape = infer_shape(&json!([1, 2, 3]));
        assert_eq!(types::ARRAY, shape.type_);
        assert_eq!(types::INTEGER, shape.array.tuple[0].type_);

        let shape = infer_shape(&json!([[1], [2.5], ["3"]]));
        assert_eq!(types::ARRAY, shape.type_);
        assert_eq!(types::ARRAY, shape.array.tuple[0].type_);
        assert_eq!(
            types::INT_OR_FRAC | types::STRING,
            shape.array.tuple[0].array.tuple[0].type_
        );

        let shape = infer_shape(&json!([[["3 layers deep"]]]));
        assert_eq!(
            types::STRING,
            shape
                // one
                .array
                .tuple[0]
                // two
                .array
                .tuple[0]
                // three
                .array
                .tuple[0]
                // type of items
                .type_
        );
    }

    #[test]
    fn build_object_types() {
        let shape = infer_shape(&json!({"a": true, "b": null, "c": 3}));
        assert_eq!(types::OBJECT, shape.type_);
        let a = get(&shape.object, "a");
        assert_eq!(types::BOOLEAN, a.shape.type_);
        let b = get(&shape.object, "b");
        assert_eq!(types::NULL, b.shape.type_);
        let c = get(&shape.object, "c");
        assert_eq!(types::INTEGER, c.shape.type_);

        let shape = infer_shape(&json!({"a": {"b": {"c": 3}}}));
        assert_eq!(types::OBJECT, shape.type_);
        let a = get(&shape.object, "a");
        assert_eq!(types::OBJECT, a.shape.type_);
        let b = get(&a.shape.object, "b");
        assert_eq!(types::OBJECT, b.shape.type_);
        let c = get(&b.shape.object, "c");
        assert_eq!(types::INTEGER, c.shape.type_);
    }

    fn get<'o>(object: &'o ObjShape, prop_name: &str) -> &'o ObjProperty {
        object
            .properties
            .iter()
            .find(|p| p.name == prop_name)
            .expect("key to exist")
    }
}
