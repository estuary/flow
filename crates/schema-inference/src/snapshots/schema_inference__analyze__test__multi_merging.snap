---
source: crates/schema-inference/src/analyze.rs
assertion_line: 158
expression: schema
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type": "object",
  "properties": {
    "a": {
      "type": [
        "integer",
        "object"
      ],
      "properties": {
        "imaginary": {
          "type": "integer"
        },
        "real": {
          "type": "number"
        }
      }
    },
    "b": {
      "type": "array",
      "items": {
        "type": [
          "boolean",
          "object"
        ],
        "properties": {
          "harmless": {
            "type": "string"
          },
          "hurtful": {
            "type": "boolean"
          },
          "truthful": {
            "type": [
              "boolean",
              "string"
            ]
          }
        }
      }
    },
    "c": {
      "type": "object",
      "properties": {
        "even_more_optional": {
          "type": "boolean"
        },
        "inner": {
          "type": [
            "null",
            "object",
            "string"
          ],
          "properties": {
            "details": {
              "type": "string"
            }
          }
        },
        "optional": {
          "type": "array",
          "items": {
            "type": "integer"
          }
        }
      }
    },
    "x": {
      "type": "string"
    },
    "y": {
      "type": "array",
      "items": {
        "type": "string"
      }
    },
    "z": {
      "type": "null"
    }
  }
}
